src/mongocxx/collection.cpp

```diff
@@ -431,12 +431,85 @@ std::int64_t collection::count(bsoncxx::document::view filter, const options::co
 }
 
 bsoncxx::document::value collection::create_index(bsoncxx::document::view keys,
-                                                  bsoncxx::document::view options) {
+                                                  const options::index& options) {
```

Comment:
So this change touches on two fairly fundamental issues that we still need to discuss. The nice thing about taking a view for the document options is that it is 100% flexible. If new index options come into being, the API doesn't need to change. That gives us great ABI stability, but a poor interface.

On the other hand, the new options::index struct gives us a very nice interface, but locks us into a strict ABI. 

And, this is not the only place we face this. Have a look at https://jira.mongodb.org/browse/CXX-435

I think there are a few ways forward that we could consider:

- For every place that we take an option type, also provide an overload that takes a document::view. Then, make every option type have a to_bson method. That way, you can use the helper to build up as much as you can, then make any customizations, and call the ABI safe method.
- Alternatively, make every option type have an "extensions" member, that takes a document_view and which it will fold in to the options set via the option type.

This has been in the back of my mind for a while, but it seems like a good time to start talking about it.